graphdatascience/query_runner/protocol/project_protocols.py

```diff
@@ -0,0 +1,88 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional
+
+from pandas import DataFrame
+
+from graphdatascience import QueryRunner
+from graphdatascience.call_parameters import CallParameters
+from graphdatascience.session.dbms.protocol_version import ProtocolVersion
+
+
+class ProjectProtocol(ABC):
+    @abstractmethod
+    def project_params(
+        self, graph_name: str, query: str, params: Dict[str, Any], arrow_config: Dict[str, Any]
+    ) -> CallParameters:
+        """Transforms the given parameters into CallParameters that correspond to the right protocol version."""
+        pass
+
+    @abstractmethod
+    def run_projection(
+        self,
+        query_runner: QueryRunner,
+        endpoint: str,
+        params: CallParameters,
+        yields: Optional[List[str]] = None,
+        database: Optional[str] = None,
+        logging: bool = False,
+    ) -> DataFrame:
+        """Returns the procedure name for the corresponding protocol version."""
+        pass
+
+    @staticmethod
+    def select(protocol_version: ProtocolVersion) -> "ProjectProtocol":
+        return {ProtocolVersion.V1: ProjectProtocolV1(), ProtocolVersion.V2: ProjectProtocolV2()}[protocol_version]
```

Comment:
I think I would prefer a match statement here. Not sure how clever the type checker is but maybe this would then tell us if we forgot a version?

Reply:
I wanted to use that, but `match` is only supported from python 3.10 on. We currently require python `>=3.8`

Reply:
Argh damn it :D